---
title: "Example code for single-cell droplet quality control"
author: "Kevin Rue-Albrecht"
date: "05/10/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(DropletUtils)
library(tidyverse)
library(ggplot2)
library(cowplot)
library(scDblFinder)
library(Seurat)
library(DoubletFinder)
library(celda)
library(scater)
```

# Exercise

## Empty droplets

- Import the raw matrix into R; use `DropletUtils`.
  Make a `SingleCellExperiment` object called `sce`.

```{r}
library(DropletUtils)
setwd("/project/obds/hfuchs/4_scRNAseq/4_singlecell_droplets")
sce <- read10xCounts(samples = c("pbmc5k" = "raw_feature_bc_matrix"), 
                     col.names = TRUE)
sce
```

- Remove droplets without any count at all.

This will massively speed up downstream steps, while removing unquestionable empty droplets.

```{r}
sce <- sce[,colSums(counts(sce)) > 0]
min(colSums(counts(sce)))
sce
```

- Run `DropletUtils::emptyDrops()`.

```{r}
set.seed(1)
out <- DropletUtils::emptyDrops(sce,
                                assay.type = "counts", 
                                lower = 100, #lower cutoff of cells at which to start calculating rather than automatically assuming negative 
                                niters = 10000)
out 
out %>% as.data.frame() %>% filter(Limited == FALSE) %>% arrange(desc(Total))
out %>% as.data.frame() %>% filter(FDR < 0.01)
```

**How do you read the output? Which droplets would you consider empty?**

> Answer:
>First column: total number of counts - any less than a hundred don't even bother with calculations 
>larger count cells: perform the calculations - FDR is the final value 
>FDR aka Q value (type one error) - more robust as multiple test correction (Benjamin-Hochberg correction)
> 

- Remove empty droplets from the data sets.
  Call the new object `sce_filtered_emptydrops`.

```{r}
sce_filtered_emptydrops <- sce[,which(out$FDR < 0.01)]
sce_filtered_emptydrops
```

## Bonus point

- Draw plots comparing empty droplets and other droplets
  (e.g., library size, mitochondrial content, PCA).

```{r}
plot_data <- tibble(Barcode = colData(sce)$Barcode, 
                    Library_Size = colSums(assay(sce, "counts")), 
                    Droplet_Status = (out$FDR < 0.01))
plot_data$Droplet_Status[is.na(plot_data$Droplet_Status) == TRUE] <- "Assigned_Empty"
levels(plot_data$Droplet_Status) <- c("Assigned_Empty","FALSE","TRUE")
plot_data 

ggplot(plot_data,aes(x = Droplet_Status, y = Library_Size)) +
  geom_violin() + 
  scale_y_log10() +
  theme_minimal()

#alogirthm calculates average gene expressison profile of empty droplets, from then assigns cells
#more sophisticated than just counting 
  
```

```{r}
is.mt <- 
# table(is.mt)
plot_data <- tibble(
  
  
  
  
)
ggplot(   ) +
  
  
  
```

```{r}

sce <- 
sce <- 
plot_data <- tibble(
  
  
)
ggplot(   ) +
  
  
```

- Import the filtered matrix for the same cells and compare their set of filtered cells to yours.

```{r}
sce_filtered_cellranger <- 
  
  


all_barcodes <- colnames(sce)

compare_table <- tibble(
  emptydrops = all_barcodes %in%    ,
  cellranger = all_barcodes %in%    ,
)
table(compare_table)
```

# Exercise

## Doublets

Continue using the data set after you excluded empty droplets.

- Run `scDblFinder::scDblFinder()`.

```{r}
library(scDblFinder)
set.seed(1)
sce_scdblfinder <- scDblFinder(sce_filtered_emptydrops)
table(sce_scdblfinder$scDblFinder.class)
```

**How do you read the output? Which droplets would you consider doublets?**

> Answer:
> Droples have high/low scDblFinder score or scDblFinder.class = "doublet"

- Remove doublets from the data set.
  Call the new object `sce_filtered_scdblfinder`.

```{r}
sce_filtered_scdblfinder <- sce_scdblfinder[,sce_scdblfinder$scDblFinder.class == "singlet"]
sce_filtered_scdblfinder
```

## Bonus point

- Draw plots comparing doublets and other droplets.
  (e.g., library size, mitochondrial content, PCA)

```{r}
is.mt <- grep("^MT-", rowData(sce_scdblfinder)$Symbol) 
table(is.mt)
library(scuttle)
sce_scdblfinder <- scuttle::addPerCellQC(sce_scdblfinder, percent_top = 50, subsets = list(MT=is.mt))

plot_data <- tibble(
  Doublet_ID = sce_scdblfinder$scDblFinder.class,
  Library_size = colSums(assay(sce_scdblfinder, "counts")), 
  Mitochondrial = colData(sce_scdblfinder)$subsets_MT_percent)
ggplot(plot_data, aes(x = Doublet_ID, y = Mitochondrial)) + geom_violin()
ggplot(plot_data, aes(x = Doublet_ID, y = Library_size)) + geom_violin()

```
```{r}
sce_scdblfinder <- logNormCounts(sce_scdblfinder)
set.seed(1)
sce_scdblfinder <- runPCA(sce_scdblfinder)






#UMAP
sce_scdblfinder <- runUMAP(sce_scdblfinder, dimred = "PCA", n_dimred = 1:10)

plot_data <- tibble(
  barcode = colnames(sce_scdblfinder),
  UMAP1 = reducedDim(sce_scdblfinder, "UMAP")[,1],
  UMAP2 = reducedDim(sce_scdblfinder, "UMAP")[,2],
  droplet_type = sce_scdblfinder$scDblFinder.class
)

ggplot() +
  geom_point(data = plot_data %>% filter(droplet_type == "singlet"),
             mapping = aes(x = UMAP1, y = UMAP2), size = 0.5, colour = "black") +
  geom_point(data = plot_data %>% filter(droplet_type == "doublet"),
             mapping = aes(x = UMAP1, y = UMAP2), size = 0.5, colour = "red") + 
  theme_cowplot()
  
  
```

- Compare results with the output of `chris-mcginnis-ucsf/DoubletFinder`

  + You will need to make a `Seurat` object.

```{r, message=FALSE}
## Pre-process Seurat object (standard)
seu <- 
seu <- 
seu <- 
seu <- 
seu <- 
seu <- 
```

```{r, message=FALSE}
## pK Identification (no ground-truth) 

sweep.res.list <- 

sweep.stats <- 
bcmvn <- 
```

```{r}
ggplot() +
  
  
  
```

```{r}
## Homotypic Doublet Proportion Estimate 
homotypic.prop <- 
nExp_poi <- 
nExp_poi.adj <- 
```

```{r, message=FALSE}
## Run DoubletFinder with varying classification stringencies 
seu <- 
```

```{r}
table(   )
```

```{r, message=FALSE}
seu <- 
```

```{r}
table(   )
```

```{r}
doublets_table <- data.frame(
  
  
  
)
table(doublets_table)
```

# Exercise

## Ambient RNA

- Run `celda::decontX()` (continue using the data after excluding empty droplets and doublets).

```{r, message=FALSE}
library(celda)
set.seed(1)
sce_decontX <- decontX(sce_filtered_scdblfinder)

colData(sce_decontX) %>% head()# $decontX_contamination (score) decontX_clusters (ID)
rowData(sce_decontX)


```

- Visualize the UMAP layout created by `decontX`; use `plotDimReduceCluster()`.

```{r}
plotDimReduceCluster(colData(sce_decontX)$decontX_clusters,
                     dim1 =  reducedDim(sce_decontX)[,1], #package-calculated UMAP 
                     dim2 = reducedDim(sce_decontX)[,2]) #need to manually tell it where to find it

#this is just a qc plot telling us what the method uses as clusters
#if this doesn't look sensible, might have to provide own clustering etc.
```

- Visualize the percentage of contamination estimated by `decontX` in each cell.
  Use the `plotDecontXContamination()` function.

```{r}
plotDecontXContamination(sce_decontX)
```

**Does any cluster display higher (estimated) levels of ambient RNA?**

> Answer: cluster 4
> 
> 

- Visualize the log-normalised expression of marker genes on a UMAP layout faceted by gene; use `plotDimReduceFeature()`.

**Note:** For instance, use `CD3D`, `CD3E`, `GNLY`, `LYZ`, `S100A8`, `S100A9`, `CD79A`, `CD79B`, and `MS4A1`.

<!--
Source: https://bioconductor.org/packages/devel/bioc/vignettes/celda/inst/doc/decontX.pdf
(Section 5.3)
We will use:
- CD3D and CD3E for T-cells,
- LYZ, S100A8, and S100A9 for monocytes,
- CD79A, CD79B, and MS4A1 for B-cells,
- GNLY for NK-cells,
- and PPBP for megakaryocytes
-->

```{r}

features_of_interest <- c("CD3D", "CD3E", "GNLY", "LYZ", "S100A8", "S100A9", "CD79A", "CD79B", "PPBP")
#use gene symbol if unique, append EnsemblID if necessary (could also manually change that)
#more sophisticated value of make.unique 
rownames(sce_decontX) <- uniquifyFeatureNames(rowData(sce_decontX)$ID, rowData(sce_decontX)$Symbol)
plotDimReduceFeature(assay(sce_decontX,"decontXcounts"), 
                     dim1 = reducedDim(sce_decontX)[,1],
                     dim2 = reducedDim(sce_decontX)[,2],
                     features = features_of_interest, 
                     useAssay = "decontXcounts")
```

- Visualise the percentage of cells with detectable expression of marker genes before and after decontamination as a bar plot; use `plotDecontXMarkerPercentage()`.

```{r}
markers <- list(Tcell_Markers = c("CD3D", "CD3E"),
                NK_Markers = c("GNLY"),
                Monocyte_Markers = c("LYZ", "S100A8", "S100A9"),
                Bcell_Markers = c("CD79A", "CD79B"),
                Mega_Markers = c("PPBP"))
cellTypeMappings <- list(Tcells = 1,
                NK = 5,
                Monocytes = 2,
                Bcells = 3,
                Megas = 7)
p1 <- plotDecontXMarkerPercentage(sce_decontX, 
                                  groupClusters = cellTypeMappings,
                            markers = markers, 
                            assayName = "counts")

p2 <- plotDecontXMarkerPercentage(sce_decontX, 
                                  groupClusters = cellTypeMappings,
                            markers = markers, 
                            assayName = "decontXcounts")

plot_grid(p1,p2)
```

- Visualize the log-normalised expression of marker genes before and after decontamination as a violin plot; use `plotDecontXMarkerExpression()`.

```{r, fig.height = 12, fig.width = 12}
plotDecontXMarkerExpression(sce_decontX,
                            markers = features_of_interest,
                            groupClusters = cellTypeMappings,
                            assayName = c("counts","decontXcounts"))

```
